Implement programmatic pipeline features

weierophinney opened the pull request:
This patch implements the [Programmatic Pipelines RFC requirements](https://gist.github.com/weierophinney/a983f733c2ff3c9b3a687625e012e6e1#requirements-for-expressive-11).

TODO
----

- [x] Create `Zend\Expressive\Middleware\ErrorResponseGenerator`
- [x] Create `Zend\Expressive\Middleware\WhoopsErrorResponseGenerator`
- [x] Create `Zend\Expressive\Middleware\NotFoundHandler`
- [x] Create `Zend\Expressive\Container\ErrorResponseGeneratorFactory`
- [x] Create `Zend\Expressive\Container\WhoopsErrorResponseGeneratorFactory`
- [x] Create `Zend\Expressive\Container\NotFoundHandlerFactory`
- [x] Create `Zend\Expressive\Container\ErrorHandlerFactory`
- [x] Create `Zend\Expressive\ApplicationConfigInjectionTrait`, and compose it into `Application`.
- [x] Update `Zend\Expressive\Container\ApplicationFactory` to vary creation and return of the `Application` instance based on the `zend-expressive.programmatic_pipeline` configuration flag.
- [x] Update `Zend\Expressive\Container\ApplicationFactory` to use the methods defined in `ApplicationConfigInjectionTrait` to inject pipeline and routed middleware from configuration into the `Application` instance.
- [x] Update `Zend\Expressive\Application` to emit a deprecation notice from `pipeErrorMiddleware()`.
- [x] Update `Zend\Expressive\MarshalMiddlewareTrait` to allow marshalling http-interop middleware.
- [x] Add support for "raise throwables" flag to `ApplicationFactory`.
- [x] Documentation

## Tooling support

The following tools were written in [zendframework/zend-expressive-tooling](https://github.com/zendframework/zend-expressive-tooling) to support user migrations to work with the above changes:

- [x] Tool for generating programmatic pipelines and routing from existing configuration.
- [x] Tool for migrating `getOriginal*()` calls to `getAttribute(*)` calls on the request object.
- [x] Tool for identifying error middleware in an existing application.

References
----------

- [Programmatic Pipelines RFC](https://gist.github.com/weierophinney/a983f733c2ff3c9b3a687625e012e6e1)
- #392 (original draft of error handler classes)
- [Stratigility 1.3/2.0 migration docs](https://docs.zendframework.com/zend-stratigility/migration/to-v2/)

weierophinney commented:
Ready for review! Pinging:

- @zendframework/community-review-team 
- @michaelmoussa 
- @xtreamwayz 
- @ezimuel 

Thanks in advance!

Koopzington commented on composer.json:
```diff
@@ -52,15 +54,18 @@
-        "cs-fix": "phpcbf",
-        "test": "phpunit",
+        "cs-check": "phpcs --colors",
+        "cs-fix": "phpcbf --colors",
```
--colors for phpcs and phpcbf is not needed since the new coding standard will have them enabled via the ruleset.xml

xtreamwayz commented on doc/book/reference/migration/to-v1.1.md:
```diff
@@ -0,0 +1,560 @@
+- `delete($path, $middleware, $name = null)`
+- `route($path, $middleware, array $methods = null, $name = null)`
+
+Each returns a `Zend\Expressive\Router\Route` instance; this is useful if you
```
Thanx for mentioning this. No need anymore for routes via config.

weierophinney replied:
This has actually been true all along... just not well documented.

xtreamwayz commented on doc/book/reference/migration/to-v1.1.md:
```diff
@@ -0,0 +1,560 @@
+- Decorate the middleware in a `Zend\Stratigility\Middleware\CallableMiddlewareWrapper`
+  instance (which also requires a `$responsePrototype`).
+
+We recommend that you begin writing middleware to follow the http-interop
```
Maybe add an example here to clarify it a bit. I think it's something like this? (Disclaimer: from memory and untested)

```php
<?php

namespace App\Middleware;

use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use Interop\Http\Middleware\ServerMiddlewareInterface;
use Interop\Http\Middleware\DelegateInterface;

class MyMiddleware implements ServerMiddlewareInterface
{
    public function __construct()
    {
    }

    /**
     * @param Request           $request
     * @param DelegateInterface $delegate
     *
     * @return Response
     */
    public function process(ServerRequestInterface $request, DelegateInterface $delegate): ResponseInterface
    {
        // Do something (with the request) first

        // Call the next middleware and wait for the response
        $response = $delegate->process($request);

        // Do something (with the response) before returning the response

        // Return the response
        return $response;
    }
}
```

weierophinney replied:
Excellent feedback! I'll incorporate that shortly!

harikt commented on doc/book/reference/migration/to-v1.1.md:
```diff
@@ -0,0 +1,560 @@
+  middleware_ using the caught exception as the `$err` argument.
+
+- The "Final Handler". This is a special middleware type with the signature
+  `function ServerRequestInterface $request, ResponseInterface $response, $err = null)`,
```
Probably missed a `(`

glen-84 commented on doc/book/reference/migration/to-v1.1.md:
````diff
@@ -0,0 +1,560 @@
+
+  ```php
+  include 'config/pipeline.php';
+  include 'config/routes.php';
````
Shouldn't these use `require`?

weierophinney replied:
Yeah, probably. Thanks for the suggestion; will implement shortly.

webimpress commented on test/ApplicationConfigInjectionTest.php:
```diff
@@ -0,0 +1,414 @@
+namespace ZendTest\Expressive;
+
+use Closure;
+use InvalidArgumentException;
```
It is never used.

webimpress commented on src/ApplicationConfigInjectionTrait.php:
```diff
@@ -0,0 +1,307 @@
+
+namespace Zend\Expressive;
+
+use ReflectionProperty;
```
It is never used.

webimpress commented on src/Container/ApplicationFactory.php:
```diff
@@ -7,16 +7,16 @@
 use Zend\Diactoros\Response\EmitterInterface;
 use Zend\Expressive\Application;
+use Zend\Expressive\ApplicationUtils;
 use Zend\Expressive\Exception;
```
It is no longer used here, can be removed.

webimpress commented on test/Container/ErrorResponseGeneratorFactoryTest.php:
```diff
@@ -0,0 +1,82 @@
+use Interop\Container\ContainerInterface;
+use PHPUnit_Framework_TestCase as TestCase;
+use Zend\Expressive\Container\ErrorResponseGeneratorFactory;
+use Zend\Expressive\Middleware\ErrorResponseGenerator;
```
It is never used.

weierophinney replied:
It is - but using the `::class` notation, for `assertInstanceOf` checks.

webimpress commented on bin/expressive-pipeline-from-config:
```diff
@@ -0,0 +1,27 @@
+
+namespace Zend\Expressive;
+
+use Zend\Stdlib\ConsoleHelper;
```
Seems to be unused in this file.

webimpress commented on test/Command/GenerateProgrammaticPipelineFromConfig/GeneratorTest.php:
```diff
@@ -0,0 +1,158 @@
+use org\bovigo\vfs\vfsStream;
+use org\bovigo\vfs\vfsStreamDirectory;
+use PHPUnit_Framework_TestCase as TestCase;
+use Prophecy\Argument;
```
Unused, please remove.

webimpress commented on test/Command/GenerateProgrammaticPipelineFromConfig/GeneratorTest.php:
```diff
@@ -0,0 +1,158 @@
+namespace ZendTest\Expressive\Command\GenerateProgrammaticPipelineFromConfig;
+
+use org\bovigo\vfs\vfsStream;
+use org\bovigo\vfs\vfsStreamDirectory;
```
Unused, please remove.

weierophinney replied:
Odd, I needed to have this in another test suite whenever I used the `vfsStream::setup()` method; evidently that's no longer true in recent versions. Removing.

webimpress commented on test/Command/GenerateProgrammaticPipelineFromConfig/CommandTest.php:
```diff
@@ -0,0 +1,138 @@
+namespace ZendTest\Expressive\Command\GenerateProgrammaticPipelineFromConfig;
+
+use org\bovigo\vfs\vfsStream;
+use org\bovigo\vfs\vfsStreamDirectory;
```
Unused, please remove.

webimpress commented on src/Command/GenerateProgrammaticPipelineFromConfig/Command.php:
```diff
@@ -0,0 +1,165 @@
+    /**
+     * @param ConsoleHelper $helper
+     */
+    public function __construct($command = self::DEFAULT_COMMAND_NAME, ConsoleHelper $helper = null)
```
Missing `$command` in PHPDoc

webimpress commented on src/Command/GenerateProgrammaticPipelineFromConfig/Help.php:
```diff
@@ -0,0 +1,71 @@
+    /**
+     * @param ConsoleHelper $helper
+     */
+    public function __construct($command, ConsoleHelper $helper)
```
Missing `$command` in PHPDoc.

webimpress commented on src/Command/GenerateProgrammaticPipelineFromConfig/Help.php:
```diff
@@ -0,0 +1,71 @@
+<?php
+/**
+ * @see       https://github.com/zendframework/zend-expressive for the canonical source repository
+ * @copyright Copyright (c) 2015-2016 Zend Technologies USA Inc. (http://www.zend.com)
```
This is general one: shouldn't we just use `2016` year for new files?

webimpress commented on src/Container/NotFoundHandlerFactory.php:
```diff
@@ -0,0 +1,37 @@
+class NotFoundHandlerFactory
+{
+    /**
+     * @param ContainerInterface
```
Missing param name `$container`

webimpress replied:
Please add param name `$container` in PHPDoc after type `ContainerInterface`.

webimpress commented on src/Middleware/WhoopsErrorResponseGenerator.php:
```diff
@@ -0,0 +1,88 @@
+     * @param \Throwable|\Exception $e
+     * @param ServerRequestInterface $request
+     * @param ResponseInterface $response
+     * @return $response
```
Should be here return __type__, not a variable name.

webimpress commented on src/Container/ApplicationFactory.php:
```diff
@@ -7,16 +7,16 @@
-use SplPriorityQueue;
 use Zend\Diactoros\Response\EmitterInterface;
 use Zend\Expressive\Application;
+use Zend\Expressive\ApplicationUtils;
```
I can't see this class in the library. Where is it?

webimpress commented on src/Container/ApplicationFactory.php:
```diff
@@ -27,105 +27,41 @@
  *
- * When introspecting the `config` service, the following structure can be used
- * to define routes:
+ * Please see the `Zend\Expressive\ApplicationUtils` class for details on how
```
As above, I can't see this class in the library. Am I missing something?

webimpress commented on src/Command/GenerateProgrammaticPipelineFromConfig/Generator.php:
```diff
@@ -0,0 +1,458 @@
+     */
+    private function createOptionValue($value, $indentLevel = 1)
+    {
+        if (is_array($value) || $value instanceof Traversable) {
```
`Traversable` should be imported or `\Traversable` here.

weierophinney replied:
Done!

webimpress commented on src/Middleware/WhoopsErrorResponseGenerator.php:
```diff
@@ -0,0 +1,88 @@
+class WhoopsErrorResponseGenerator
+{
+    /**
+     * @var Whoops
```
Seems to be wrong type here.

webimpress commented on src/Command/GenerateProgrammaticPipelineFromConfig/Generator.php:
```diff
@@ -0,0 +1,458 @@
+     * @param int $indentLevel
+     * @return string
+     */
+    private function formatOptions($options, $indentLevel = 1)
```
Is `$options` arg always an array? Can we add type hint here?

weierophinney replied:
It could be an `ArrayObject`.

webimpress replied:
So then PHPDoc should be updated.

webimpress commented on doc/book/reference/migration/to-v1.1.md:
```diff
@@ -0,0 +1,560 @@
+## http-interop
+
+Stratigility 1.3 provides the ability to work with [http-interop middleware
+0.2.0](https://github.com/http-interop/http-middleware/tree/ff545c87e97bf4d88f0cb7eb3e89f99aaa53d7a9).
```
Maybe better here link to tagged version:
https://github.com/http-interop/http-middleware/tree/0.2.0

webimpress commented:
In `Zend\Expressive\Application` there are some methods and properties marked as:
```
@deprecated This property will be removed in v1.1.
```

and also:

```
@todo For 1.1, remove the RouteResultSubjectInterface implementation, and
      all deprecated properties and methods.
```

Is it gonna be done in this PR, or PR https://github.com/zendframework/zend-expressive/pull/376 will be merged before release 1.1?

webimpress commented:
Migration script:
I would expect that after using migration script my application will work straight away, but I've got error:

```
Fatal error: Uncaught exception 'Zend\Expressive\Exception\InvalidMiddlewareException' with message 'Unable to resolve middleware "string[Scalar]" to a callable' in my-app\vendor\zendframework\zend-expressive\src\MarshalMiddlewareTrait.php on line 65
```

This is because migration script is not adding:
```
    'dependencies' => [
        'invokables' => [
            OriginalMessages::class => OriginalMessages::class,
```

Into configuration, but:

```
$app->pipe(\Zend\Stratigility\Middleware\OriginalMessages::class);
```
is added into `config/pipeline.php`.

I think it should be done automatically (in migration script) or it should be noted in migration notes, that this action is required to be done manually.

webimpress commented:
One more issue. This was my config based configuration - `config/autoload/middleware-pipeline.global.php`:
```php
[...]
    'middleware_pipeline' => 
        'my-path' => [
            'path' => '/my-path',
            'middleware' => [
                ApplicationFactory::ROUTING_MIDDLEWARE,
                \App\Middleware\MyMiddleware::class,
                \App\Middleware\OtherMiddleware::class,
            ],
        ],
```

it was successfully converted to middleware pipeline (via migration script) - `config/pipeline.php`:

```php
[...]
$app->pipe('/my-path', [
    \Zend\Expressive\Application::ROUTING_MIDDLEWARE,
    \App\Middleware\MyMiddleware::class,
    \App\Middleware\OtherMiddleware::class,
]);
```

And then piping is failing, because `pipe` method from Application call `prepareMiddleware` method from `Zend\Expressive\MarshalMiddlewareTrait` with `$middleware = "EXPRESSIVE_ROUTING_MIDDLEWARE"`, and we got fatal:
```
Fatal error: Uncaught exception 'Zend\Expressive\Exception\InvalidMiddlewareException' with message 'Unable to resolve middleware "string[Scalar]" to a callable' in my-app\vendor\zendframework\zend-expressive\src\MarshalMiddlewareTrait.php on line 65
```

Am I doing something wrong or there are missing some scenarios, when we have some routing/dispatch middlewares only for some paths?

webimpress commented:
Another issue. "Method not allowed". I have routing, let say:
```php
$app->post('/foo', \App\Action\Foo::class, 'foo');
```

So only POST request are allowed. If I try to do `GET /foo` I'm getting 200 response with no body.
I should get 405.

I think the problem is in `\Zend\Expressive\Application::routeMiddleware:452` where we have:
```php
        $result = $this->router->match($request);

        if ($result->isFailure()) {
            if ($result->isMethodFailure()) {
                $response = $response->withStatus(405)
                    ->withHeader('Allow', implode(',', $result->getAllowedMethods()));
                return $next($request, $response, 405); // <-------
            }
[...]
```

and then, the 3rd argument (`405`) is passed through `\Zend\Stratigility\Next:119` where `$err = 405`. Error is not thrown (because `$err` is 405 not an instance of `\Exception|\Throwable`).
I can write here more, but I think it should be enough to investigate and fix this issue.

weierophinney commented:
> Is it gonna be done in this PR, or PR #376 will be merged before release 1.1?

I'll be merging #376 before release, which covers these.

weierophinney commented:
> Another issue. "Method not allowed". I

Should be resolved now; thanks for finding the issue!

weierophinney commented:
> And then piping is failing, because pipe method from Application call prepareMiddleware method from Zend\Expressive\MarshalMiddlewareTrait with $middleware = "EXPRESSIVE_ROUTING_MIDDLEWARE", and we got fatal

I think this is an edge case we never tested.

When you use configuration-based pipelines, as it loops through nested pipelines, if it encounters either of those special constants, it replaces the middleware with the appropriate callable (`[$this, 'routeMiddleware]` and `[$this, 'dispatchMiddleware]`).

However, in the `MarshalMiddlewareTrait`, we do not.

I'm going to see if I can create a test case and resolve this; I think it should be relatively easy to accomplish.

weierophinney commented on src/Command/GenerateProgrammaticPipelineFromConfig/Generator.php:
```diff
@@ -0,0 +1,458 @@
+<?php
+/**
+ * Expressive middleware pipeline
+ */
```
Done!

danizord commented on src/Container/ErrorHandlerFactory.php:
```diff
@@ -0,0 +1,29 @@
+            ? $container->get(ErrorResponseGenerator::class)
+            : null;
+
+        return new ErrorHandler(new Response(), $generator);
```
@weierophinney any chance to make it possible to configure `ErrorHandler` listeners via config? Or should I override this factory to attach my custom listeners?

xtreamwayz commented:
@weierophinney I only see the config setting for the new error handling in the migration guide:
`'zend-expressive' => ['raise_throwables' => true,],`

Did you forgot `$app->raiseThrowables();` or is the config setting preferred?

xtreamwayz commented on doc/book/reference/migration/to-v1.1.md:
```diff
@@ -0,0 +1,604 @@
+
+Stratigility 1.3 deprecates its internal request and response decorators,
+`Zend\Stratigility\Http\Request` and `Zend\Stratigility\Http\Response`,
+respsectively. The main utility of these instances was to provide access in
```
Typo: respectively

weierophinney replied:
Fixed, thanks!

xtreamwayz commented:
Also phpunit can be upgraded to 5.x since PHP 5.5 support is removed.

weierophinney commented:
> Did you forgot $app->raiseThrowables(); or is the config setting preferred?

You can do either; I didn't document the method, as technically it's part of the Stratigility 1.3 API, and documented there.

weierophinney commented on doc/book/reference/migration/to-v1.1.md:
```diff
@@ -0,0 +1,604 @@
+programmatic/declarative statements. Specifically:
+
+- We recommend putting the pipeline declarations into `config/pipeline.php`.
+- We recommend putting the pipeline declarations into `config/routes.php`.
```
Fixed, thanks!

weierophinney commented:
> Also phpunit can be upgraded to 5.x since PHP 5.5 support is removed.

Done in latest commit; thanks!

weierophinney commented:
@webimpress On Friday, I resolved your issues with using the `ROUTING_MIDDLEWARE`/`DISPATCH_MIDDLEWARE` inside arrays of middleware when defining a programmatic pipeline; please test again!

webimpress commented on bin/expressive-migrate-original-messages:
```diff
@@ -0,0 +1,30 @@
+ * Script for migrating configuration-driven pipelines/routes to programmatic usage.
+ *
+ * @see       https://github.com/zendframework/zend-expressive for the canonical source repository
+ * @copyright Copyright (c) 2015-2016 Zend Technologies USA Inc. (http://www.zend.com)
```
Should be just `2016` in new files.

webimpress commented on bin/expressive-pipeline-from-config:
```diff
@@ -0,0 +1,30 @@
+ * Script for migrating configuration-driven pipelines/routes to programmatic usage.
+ *
+ * @see       https://github.com/zendframework/zend-expressive for the canonical source repository
+ * @copyright Copyright (c) 2015-2016 Zend Technologies USA Inc. (http://www.zend.com)
```
As above, just `2016`.

webimpress commented on src/Command/MigrateOriginalMessageCalls/Help.php:
```diff
@@ -0,0 +1,82 @@
+<?php
+/**
+ * @see       https://github.com/zendframework/zend-expressive for the canonical source repository
+ * @copyright Copyright (c) 2015-2016 Zend Technologies USA Inc. (http://www.zend.com)
```
Only `2016` in new files, please.

webimpress commented on test/Command/MigrateOriginalMessageCalls/HelpTest.php:
```diff
@@ -0,0 +1,39 @@
+<?php
+/**
+ * @see       https://github.com/zendframework/zend-expressive for the canonical source repository
+ * @copyright Copyright (c) 2015-2016 Zend Technologies USA Inc. (http://www.zend.com)
```
Only `2016` in new files, please.

webimpress commented on test/ApplicationConfigInjectionTest.php:
```diff
@@ -0,0 +1,413 @@
+                [
+                    'path' => '/',
+                    'middleware' => $middleware,
+                    'allowed_methods' => [ 'GET' ],
```
We shouldn't have spaces after opening array brackets and before closing:
`['GET']` instead of `[ 'GET' ]`

More places in this file.

webimpress commented on test/ApplicationConfigInjectionTest.php:
```diff
@@ -0,0 +1,413 @@
+
+        $pipeline1 = [ [ 'middleware' => clone $middleware ] ];
+        $pipeline2 = [ [ 'middleware' => clone $middleware ] ];
+        $pipeline3 = [ [ 'middleware' => clone $middleware ] ];
```
Again, no spaces after opening array brackets, and before closing.

webimpress commented on doc/book/reference/migration/to-v1.1.md:
```diff
@@ -0,0 +1,623 @@
+## http-interop
+
+Stratigility 1.3 provides the ability to work with [http-interop middleware
+0.2.0](https://github.com/http-interop/http-middleware/tree/0.2.0).
```
Please note that 0.2.1 is released with REMOVED `MiddlewareInterface`, there is still `ServerMiddlewareInterface`, but I think this (remove) shouldn't be done in 0.2.X release...

webimpress commented on src/Command/MigrateOriginalMessageCalls/ConvertOriginalMessageCalls.php:
```diff
@@ -0,0 +1,150 @@
+<?php
+/**
+ * @see       https://github.com/zendframework/zend-expressive for the canonical source repository
+ * @copyright Copyright (c) 2016-2016 Zend Technologies USA Inc. (http://www.zend.com)
```
Please fix year. Should be only `2016`, the same in some other new files (12 times)

webimpress commented on src/Command/MigrateOriginalMessageCalls/ConvertOriginalMessageCalls.php:
```diff
@@ -0,0 +1,150 @@
+    /**
+     * Process a directory
+     */
+    public function process($directory)
```
Missing `$directory` param in PHPDoc.

webimpress commented on src/Command/MigrateOriginalMessageCalls/Help.php:
```diff
@@ -0,0 +1,82 @@
+    /**
+     * @param ConsoleHelper $helper
+     */
+    public function __construct($command, ConsoleHelper $helper)
```
Missing param `$command` in PHPDoc.

webimpress commented on src/Container/ApplicationFactory.php:
```diff
@@ -7,16 +7,15 @@
-use SplPriorityQueue;
 use Zend\Diactoros\Response\EmitterInterface;
 use Zend\Expressive\Application;
 use Zend\Expressive\Exception;
```
`Zend\Expressive\Exception` seems to be no longer used and can be removed.

webimpress commented on src/Container/WhoopsErrorResponseGeneratorFactory.php:
```diff
@@ -0,0 +1,25 @@
+    public function __invoke(ContainerInterface $container)
+    {
+        return new WhoopsErrorResponseGenerator(
+            $container->get('Zend\Expressive\Whoops')
```
Should we use `::class` here?

weierophinney replied:
No, because it's a virtual (vs defined) class name, and resolves to a `Whoops\Run` instance. Having it as a string helps signal that fact.

webimpress commented on bin/expressive-scan-for-error-middleware:
```diff
@@ -0,0 +1,30 @@
+ * Script for migrating configuration-driven pipelines/routes to programmatic usage.
+ *
+ * @see       https://github.com/zendframework/zend-expressive for the canonical source repository
+ * @copyright Copyright (c) 2015-2016 Zend Technologies USA Inc. (http://www.zend.com)
```
As it is a new file, there should be just `2016` year.

webimpress commented on src/Command/ScanForErrorMiddleware/ErrorMiddlewareFilter.php:
```diff
@@ -0,0 +1,123 @@
+        );
+    }
+
+    private function scanClass(SplFileInfo $file)
```
Missing PHPDoc.

webimpress commented on src/Command/ScanForErrorMiddleware/Scanner.php:
```diff
@@ -0,0 +1,129 @@
+    /**
+     * @param string path
+     * @param ConsoleHelper $console
+     * @throws ScannerException if the path is not a directory or not readable.
```
There is no any throws in construct.

webimpress commented on composer.json:
```diff
@@ -27,14 +27,15 @@
+        "mikey179/vfsStream": "^1.6",
         "mockery/mockery": "^0.9.5",
         "phpunit/phpunit": "^5.0",
+        "zendframework/zend-code": "^3.1",
```
New script "expressive-scan-for-error-middleware" requires `zend-code` to work, but this is only in `require-dev` section. So to use the script in project it's required to install `zend-code`. I think it should be noted in documentation, because this tool is not gonna work out of the box.

webimpress commented on src/Command/ScanForErrorMiddleware/Scanner.php:
```diff
@@ -0,0 +1,129 @@
+    private $path;
+
+    /**
+     * @param string path
```
Missing `$` before variable name (`path`).

webimpress commented on src/Command/ScanForErrorMiddleware/Scanner.php:
```diff
@@ -0,0 +1,129 @@
+    /**
+     * Recursively scan the given path for error middleware.
+     *
+     * @return void
```
It will be nice to keep consistency. For some methods `return void` is missing.

weierophinney replied:
Fixed.

webimpress commented on src/Command/ScanForErrorMiddleware/ErrorMiddlewareFilter.php:
```diff
@@ -0,0 +1,123 @@
+     */
+    private function callsNextWithError(ClassScanner $class)
+    {
+        return array_reduce($class->getMethods(), function ($found, $method) {
```
Can we add PHPDoc for params `$found` and `$methods`?

webimpress commented on src/Command/ScanForErrorMiddleware/ErrorMiddlewareFilter.php:
```diff
@@ -0,0 +1,123 @@
+        $file->invokableErrorMiddleware = false;
+        $file->callsNextWithError       = false;
+
+        return function ($found, $class) use ($file) {
```
Can we add PHPDoc for params `$found` and `$class`?

webimpress commented on src/Command/ScanForErrorMiddleware/NextInvocationScanner.php:
```diff
@@ -0,0 +1,120 @@
+
+class NextInvocationScanner
+{
+    const TEMPLATE = <<< 'EOT'
```
I've just checked why single quotes here:
http://php.net/manual/en/language.types.string.php#language.types.string.syntax.nowdoc
I didn't know about it. Nice 👍

webimpress commented on src/Command/ScanForErrorMiddleware/NextInvocationScanner.php:
```diff
@@ -0,0 +1,120 @@
+
+            if ($token === '(') {
+                // Hit an opening paren; increase paren depth
+                $parenDepth += 1;
```
Maybe just `++`?

weierophinney replied:
I've been using `+=` and `-=` for a few years, since seeing a Douglas Crockford talk where he demonstrated how it's more optimized both for the language interpreter (while he was talking about JavaScript, the optimizations are present for most C-style languages), as well as human comprehension (as `++` and `--` vary behavior based on whether or not they prefix or suffix the variable). Using the `+=` and `-=` syntax makes it clear that we are _re-assigning_ the value following the increment/decrement operation.

It essentially comes down to author preference; I've not been changing them during code reviews unless doing so aids in understanding how the code works.

webimpress commented on src/Command/ScanForErrorMiddleware/NextInvocationScanner.php:
```diff
@@ -0,0 +1,120 @@
+
+            if ($token === ')') {
+                // Hit an closing paren; decrease paren depth
+                $parenDepth -= 1;
```
And here `--`?

weierophinney commented:
I've just tried running each of the scripts against an older version of my own website to see how the migration would work.

The migration script for converting to a programmatic pipeline _just worked_, and my application was able to work immediately, with a few exceptions (a few deprecation notices, and places where I was using the legacy error handling).

The script for updating calls to `getOriginal*()` worked, fixing several `getOriginalRequest()` and `getOriginalUri()` calls. It also flagged one `getOriginalResponse()` call, which I was able to update to pull out of the request attribute. Again, after these changes, everything continued to _just work_.

Finally, I ran the script for identifying error middleware and/or invocation. I had a few false positives, primarily in classes implementing `__invoke()` with more than 3 arguments (delegator factories, in particular!), but it otherwise found them. This particular step is more difficult, as anything flagged needs to be manually updated. My workflow was:

- Open file, locate invocation(s).
- Determine if this is a generic error, a 404, or an app-specific error.
  - For generic errors, I simply threw a `RuntimeException`, including a status code if one was originaly given.
  - For 404 errors, I updated the middleware to compose the `NotFoundHandler`, and returned the invocation of that (see more below).
  - I had one app-specific error, for authorization. For this, I did two thingS:
    - Created a custom exception
    - Rewrote my `Unauthorized` error middleware to be standard middleware which wrapped a call to `$next` in a try/catch block, returning a response if one was returned. If the new exception was caught, I continued with my original middleware logic; for any other exception, I re-threw, so the standard error handler would catch it.

As noted above, I discovered a new use case for the `NotFoundHandler`: injecting it in middleware that may need to return a 404 response! Previously, I had several middleware that invoked `$next` with a `404` error argument, which my own final handler was picking up; now, I have these same middleware compose a `ServerMiddlewareInterface` instance that is intended to generate and return a 404 response; my factories then inject the `NotFoundHandler` for this argument.

I'm going to also try this on the example application @ezimuel used for a recent workshop, but I feel that my own website was sufficiently complex enough to demonstrate that the migration tools definitely assist in updating an application to switch to programmatic pipelines, switch to the new error handling, and help developers update their apps to remove deprecated functionality.

webimpress commented on src/Container/ApplicationFactory.php:
```diff
@@ -136,352 +70,75 @@ class ApplicationFactory
-            ? $container->get('Zend\Expressive\FinalHandler')
-            : $this->marshalFinalHandler($container);
+        $finalHandler = (isset($config['zend-expressive']['raise_throwables'])
+            && $config['zend-expressive']['raise_throwables'])
```
`! empty` should be enough here

webimpress commented on src/Container/ApplicationFactory.php:
```diff
@@ -136,352 +70,75 @@ class ApplicationFactory
-            }
-            $this->injectMiddleware($pipeline['pre_routing'], $app);
+        if (isset($config['zend-expressive']['raise_throwables'])
+            && $config['zend-expressive']['raise_throwables']
```
As above - `! empty`

webimpress commented on src/Container/ApplicationFactory.php:
```diff
@@ -136,352 +70,75 @@ class ApplicationFactory
-            }
-            $this->injectMiddleware($pipeline['post_routing'], $app);
+        if (! isset($config['zend-expressive']['programmatic_pipeline'])
+            || ! $config['zend-expressive']['programmatic_pipeline']
```
`empty` here

webimpress commented on test/Command/GenerateProgrammaticPipelineFromConfig/GeneratorTest.php:
```diff
@@ -0,0 +1,156 @@
+        return [
+            'int'        => [sprintf($pipelineConfigTemplate, '1')],
+            'float'      => [sprintf($pipelineConfigTemplate, '1.1')],
+            'object'     => [sprintf($pipelineConfigTemplate, '(object) [\'foo\' => \'bar\']')],
```
Can we have less spaces here before `=>`, please?

webimpress commented on test/Command/GenerateProgrammaticPipelineFromConfig/GeneratorTest.php:
```diff
@@ -0,0 +1,156 @@
+    'middleware_pipeline' => [
+        [
+            'middleware' => %s,
+        ]
```
Would be nice to have also here trailing comma.

webimpress commented on src/Command/MigrateOriginalMessageCalls/ConvertOriginalMessageCalls.php:
```diff
@@ -0,0 +1,153 @@
+    /**
+     * Process a single file
+     *
+     * @param $filename
```
Missing parameter type.

webimpress commented on src/Middleware/ErrorResponseGenerator.php:
```diff
@@ -0,0 +1,145 @@
+    /**
+     * @var bool $isDevelopmentMode
+     * @var null|TemplateRendererInterface $renderer
+     * @var string $template
```
`@param` instead of `@var` please

webimpress commented:
I've tested it on my application and here are my notes:

- [ ] as I've noted [before](https://github.com/zendframework/zend-expressive/pull/396#discussion_r89659704) to use "expressive-scan-for-error-middleware" script requires `zend-code` to work. It is required but in __dev__ section of composer.json and I'm not using it in my project, so I have got fatal:
  ```
  PHP Fatal error:  Class 'Zend\Code\Scanner\FileScanner' not found in C:\path-to-my-project\vendor\zendframework\zend-expressive\src\Command\ScanForErrorMiddleware\ErrorMiddlewareFilter.php on line 35
  ```
- [ ] Scripts on Windows show in help absolute path in Usage, for example:
  ```bash
  $ vendor/bin/expressive-migrate-original-messages
  Usage:

    C:/path-to-my-project/vendor/zendframework/zend-expressive/bin/expressive-migrate-original-messages [command] [options]

  Commands:
  ...
  ```
  Would be nice to have there just `vendor/bin/expressive-migrate-original-messages` instead. I don't know how it is on linux/osx.
- [ ] Consistent messages in scripts. Here we are current results (when nothing is changed):
  ```bash
  $ vendor/bin/expressive-scan-for-error-middleware scan
  Scanning for error middleware or error middleware invocation...
  Done!
  ```
  ```bash
  $ vendor/bin/expressive-migrate-original-messages scan
  // <!---------- MISSING short description here
  Done!
  ```
  If nothing is changed we see message "Done!", and I'm not sure if it is clear enough that nothing has changed. Not important, but it could improve user experience.
- [ ] Should be allowed to run "expressive-pipeline-from-config" script just once. I mean the next run should be blocked or clever detected. Files `config/pipeline.php` and `config/routes.php` can be already there, and maybe we shouldn't force override them. Also when we run it again we have got duplicated entries in `public/index.php`: `error_reporting(error_reporting() & ~E_USER_DEPRECATED);` and `require 'config/pipeline.php'; require 'config/routes.php';`, then we have got fatal:
  ```
  Fatal error: Uncaught exception 'Zend\Expressive\Exception\DuplicateRouteException' with message 'Duplicate route detected; same name or path, and one or more HTTP methods intersect'
  ```
   I know, the script is designed to run just ONCE, so maybe it's not important.
- [ ] It seems we don't need `Zend\Stratigility\Middleware\OriginalMessages` middleware right now. The functionality it delivers is duplicated, because:
  - [`Application:run()` inject `originalResponse`](https://github.com/weierophinney/zend-expressive/blob/feature/programmatic/src/Application.php#L610),
  - [`MiddlewarePipe::__invoke()` decorate request](https://github.com/zendframework/zend-stratigility/blob/master/src/MiddlewarePipe.php#L91) and [inject `originalUri` and `originalRequest` attributes](https://github.com/zendframework/zend-stratigility/blob/master/src/MiddlewarePipe.php#L254-L261) - so request is _always_ instance of `Zend\Stratigility\Http\Request` (deprecated) and request has always `original*` attributes (without using `OriginalMessages` middleware). So right now, with Stratigility 1.3, we don't need to add `OriginalMessages` middleware into our pipeline.


I tried to migrate my application using notes from [previous @weierophinney comment](https://github.com/zendframework/zend-expressive/pull/396#issuecomment-264991054) and I'm a bit confused. I'd like to use single-pass middlewares. My application checks if user is authorized at the beginning (routing middleware) - based on headers/request. This is the only place where app checks for permissions. If user is not authorized should I:
  - break pipeline and just `return new Response()->withStatusCode(403)`?
  - throw exception and catch it later and return 403?
  - pass `UnauthorizedHandler` to the middleware and process it (the same approach as suggested for `NotFoundHandler` in above mwop comment)?

If user is authorized my application returns usually `204` (no content), `200`/`201` (json responses - create/update),  `404` (not found) - it's kinda REST api. How then in single-pass approach I should return these responses:
  - pass handlers for each type of response (`NoContentHandler`, `JsonHandler`, `NotFoundHandler`...)?
  - create own responses `NoContentResponse`, `JsonResponse`, `NotFoundResponse` and just return it (break pipeline)?
  - anything else ... ?

I've got confused with single-pass approach, maybe I'm doing something wrong, or not understand it correctly. I hope it could be clarified here.

weierophinney commented:
> "expressive-scan-for-error-middleware" script requires zend-code to work.

I noticed this yesterday, but originally thought it was a fluke, so thanks for confirming! I'd forgotten that Composer no longer installs dev dependencies of dependencies when installing (it did in early versions, and we got bit by this change in other components later).

I'm not quite sure what to do about this. Expressive itself does not _require_ zend-code; only this particular script does. And I'm a bit loathe to add a _required_ dependency for something that is only to be used during _development_.

One possibility would be to create a separate package with these migration tools which users can then install as a development dependency in their projects. The problem I see with that is it's an additional step: they have to (a) read the migration documentation, (b) install the tooling, and then (c) use it, versus just (a) and (c) as we currently have. The benefit, though, is that they could then remove it when done. Additionally, they could install it globally, and run it on any project.

The only command that has any dependencies on Expressive classes is `Zend\Expressive\Command\GenerateProgrammaticPipelineFromConfig\Generator`, and this is due to constants it references. These could easily be updated to inline those values locally, which would allow the tooling to exist with only dependencies on Stratigility (for identifying ErrorMiddlewareInterface instances), zend-code, and either zend-stdlib or zf-console or symfony/console for console tooling. As a development dependency, none of these would then be installed for production deployments, which would also ensure there is no vector for abuse on production systems.

What do you think about this approach, @webimpress? /cc @zendframework/community-review-team @michaelmoussa @xtreamwayz

webimpress commented:
> One possibility would be to create a separate package with these migration tools

@weierophinney I kinda like this approach, because these tools are actually "development tools" only and we don't want them to be available in live/production environment. As you mention it will add a bit more work on migration, but as you suggest, will be possible to use it as global tools (in case someone needs migrate more projects). In expressive library we can have then entries in "suggest" and/or "conflicts" of composer.json (for example we will allow only `expressive-migration-tools ^2.0` to work with `expressive ^2.0` etc).

The more I think about it, it just seems to be right solution :)

weierophinney commented:
> Scripts on Windows show in help absolute path in Usage

In your example, I notice an additional detail: it's not the path to the symlink under `vendor/bin/`, but to the script under the package itself; it's resolving the link. (Under *nix systems, it displays `./vendor/bin/{command name}`; in other words, it's exactly what you provide on the CLI, no more, no less.)

My thought on this, then, is to do the following: if the command is a resolvable filename, and it is identical to the _realpath_ of that filename, we truncate it to only the filename. The user already has invoked it, so the full path is not necessary. If it is not a realpath, we leave it as is.

I'll make this change shortly.

weierophinney commented:
> It seems we don't need Zend\Stratigility\Middleware\OriginalMessages middleware right now.

For now, yes, but when Stratigility 2.0 is ready, it will no longer perform these injections. Having this middleware in place makes an application _future-proof_.

webimpress commented:
Maybe it's too crazy but, what if we extract migrations tools into separate library and prepare here simply _interactive migration script_, which will install composer dependencies (migrations-tool library, if needed) and display available migrations options (we can detect current version, check what is to do, etc...):
```bash
$ ./vendor/bin/expressive-migration
Installing required dependencies....

Please choose what you want to do:
  [1] migrate to pipeline from config
  [2] migrate original messages
  [3] scan for error middlewares
  [Q] quit from migration tool

Your choose ___
```

@weierophinney What do you think?
 interactive migration tool.

weierophinney commented:
@webimpress —

Regarding:

> I'd like to use single-pass middlewares. My application checks if user is authorized at the beginning (routing middleware) - based on headers/request. This is the only place where app checks for permissions. If user is not authorized should I:

and:

> How then in single-pass approach I should return these responses:

There are a few approaches to this, and you actually touch on all of them.

The reason single pass is being recommended as part of PSR-15/http-middleware is due to a pretty big issue with passing the response: if one middleware alters the response it pushes into another layer (by invoking `$next` or the delegate), there's no guarantee that those changes will be reflected in the response returned by inner layers as any of these layers could potentially create and return a brand new response. Additionally, because any given layer could write to the response body, developers need to jump a number of hurdles to ensure anything *they* write properly overwrites the original (see [Rob Allen's post on this subject](https://akrabat.com/filtering-the-psr-7-body-in-middleware/).  Anthony Ferrara has written a [comprehensive post covering these response issues](http://blog.ircmaxell.com/2016/05/all-about-middleware.html) if you want more detail.

So, knowing that passing the response via the `$next`/`$delegate` argument is not safe, how do we return a response?

You've noted several options already, and these are the suggested approaches:

- Inject an empty or pre-populated response via the middleware's constructor.  This approach allows you to typehint against PSR-7, and then use your DiC to provide the concrete implementation.

- Inject a _factory_ that is capable of producing a response via the middleware's constructor. (This is essentially the idea behind [PSR-17](https://github.com/php-fig/fig-standards/tree/master/proposed/http-factory) ([HTTP message factories](https://github.com/http-interop/http-factory))). This approach allows you to typehint against a callable or a generic interface, and, again, use your DiC to provide the concrete implementation. Such a factory could even be _other middleware_, as I noted in a previous comment with regards to re-using the `NotFoundHandler` from Expressive for generating my 404 response.

- Instantiate a concrete response within your middleware. In this case, your middleware would _depend_ on a concrete implementation of PSR-7. This is perfectly acceptable _within your application_, but for middleware you want to distribute as a library, you should likely go with the previous approaches.

I personally like the second/middle approach above a ton, as it allows me to create templated responses, or even alter responses based on `Accept` header and the like.

Now, regarding your "user not authorized" scenario, you have a couple of options:

- As noted already, you could return a response somehow, using one of the above methods. This is likely the fastest and easiest possiblity, and it's what I'm doing on the current incarnation of my website. (See [this example](https://github.com/weierophinney/mwop.net/blob/db4d5a7c1c2d2524eb7c4713f287e75e849ae5da/src/ComicsPage.php#L22-L33).) This approach is essentially the first and last bullet points you presented.

- Create a custom middleware for handling "unauthorized" exceptions, and have other middleware raise such an exception for those situations. That ends up looking like:

  ```php
  class UnauthorizedHandler
  {
      public function __invoke($request, DelegateInterface $delegate)
      {
          try {
              $response = $delegate->process($request);
              return $response;
          } catch (UnauthorizedException $e)
          } catch (\Throwable $e)
              throw $e;
          } catch (\Exception $e)
              throw $e;
          }

          // create 403 or related response and return it here...
      }
  }

  class MiddlewareRequiringAuthorization
  {
      public function __invoke($request, DelegateInterface $delegate)
      {
          if (! $this->isAuthorized($request)) {
               throw new UnauthorizedException();
          }

          // ...
      }
  }
  ```
  
It basically boils down to your own personal preference. The other thing to consider, though, is that having nests of try/catch blocks can be potentially expensive; having a lot of error handlers like the above might slow down your application. As always, test and measure to determine the impact.

weierophinney commented:
@webimpress — regarding having an interactive migration tool, that may be a bit more than I want to bite off. However, I definitely like the idea of a script that will _install_ the migration tooling; it could even detail what the newly installed tools are and their purpose following installation, which would provide the user with the information they need to proceed.

I'll work on that after separating out the tooling; thanks for the idea!

webimpress commented:
@weierophinney 2a825f4 didn't resolve the issue

```
$ composer show
zendframework/zend-expressive           dev-feature/programmatic be7224a 

$ vendor/bin/expressive-migrate-original-messages
Usage:

  C:/path-to-my-project/vendor/zendframework/zend-expressive/bin/expressive-migrate-original-messages [command] [options]
```

so still it's not working.

I debugged it and problem is with slashes/backslashes:
```
$this->command is 'C:/path-to-my-project/vendor/zendframework/zend-expressive/bin/expressive-migrate-original-messages';

file_exists($this->command) // <-- true
realpath($this->command) === $this->command // <-- false

realpath($this->command) is C:\path-to-my-project\vendor\zendframework\zend-expressive\bin\expressive-migrate-original-messages
```

weierophinney commented:
Tooling is now separated into the [zendframework/zend-expressive-tooling repository](https://github.com/zendframework/zend-expressive-tooling). I've also added a new vendor binary, `bin/expressive-tooling`, which can be used to install and remove the tooling support:

```bash
$ ./vendor/bin/expressive-tooling # installs it as a dev dependency
$ ./vendor/bin/expressive-tooling remove # removes it
```

I will update the documentation shortly to reflect these changes; otherwise, the docs are still correct in the tooling usage and what they each do.

weierophinney commented:
> I debugged it and problem is with slashes/backslashes:

If you can, could you submit a PR against the new repo to fix this? I don't have a working PHP install in any of my windows envs currently, so I cannot easily test this.

weierophinney commented on bin/expressive-tooling:
```diff
@@ -0,0 +1,90 @@
+    case 'remove':
+        echo "Removing zend-expressive-tooling\n";
+        $remove  = true;
+        $command = sprintf('%s remove --dev zendframework/zend-expressive-tooling', $composer);
```
Yes - otherwise composer will prompt you, indicating it did not find the required package, but one exists as a dev requirement, and ask if you want to remove that. Having the flag in the command allows it to run without user intervention.

weierophinney commented:
Rebased to remove commits related to the tooling support, which is now in [zendframework/zend-expressive-tooling](https://github.com/zendframework/zend-expressive-tooling). I'll add a commit shortly for the documentation to detail how to install the tooling prior to using the various scripts it defines.

weierophinney commented:
All documentation is complete; ready for final review!

malukenho commented on src/ApplicationConfigInjectionTrait.php:
```diff
@@ -0,0 +1,306 @@
+            }
+
+            if (isset($spec['allowed_methods'])) {
+                $methods = $spec['allowed_methods'];
```
why not 
```php
$methods = isset($spec['allowed_methods']) ? $spec['allowed_methods'] : Route::HTTP_METHOD_ANY;

if (! is_array($methods)) {
    throw new Container\Exception\InvalidArgumentException(sprintf(
        'Allowed HTTP methods for a route must be in form of an array; received "%s"',
        gettype($methods)
    ));
}
```

weierophinney replied:
Because `Route::HTTP_METHOD_ANY` is not an array, but an integer.

malukenho commented on src/ApplicationConfigInjectionTrait.php:
```diff
@@ -0,0 +1,306 @@
+            Application::DISPATCH_MIDDLEWARE,
+        ];
+
+        return function ($item) use ($pipelineMap, $appMiddlewares) {
```
mark as `static` to protect against `$this` access. I'm not sure if there's some gain on it on that context btw.

weierophinney replied:
Can't be static, as it's closing over provided values. Additionally, I don't see what value that would add; this is returned from a private method, so those extending would not be able to re-bind anyways.

malukenho commented on src/ApplicationConfigInjectionTrait.php:
```diff
@@ -0,0 +1,306 @@
+                throw new Container\Exception\InvalidArgumentException(sprintf(
+                    'Invalid pipeline specification received; must be an array containing a middleware '
+                    . 'key, or one of the Application::*_MIDDLEWARE constants; received %s',
+                    (is_object($item) ? get_class($item) : gettype($item))
```
unneeded `()`

weierophinney replied:
Fixed.

malukenho commented on src/MarshalMiddlewareTrait.php:
```diff
@@ -159,4 +212,46 @@ private function marshalLazyErrorMiddlewareService($middleware, ContainerInterfa
+        $r = $this->reflectMiddleware($middleware);
+        $paramsCount = $r->getNumberOfParameters();
+
+        if ($paramsCount !== 2) {
```
Can be `return $paramsCount === 2`?

weierophinney replied:
Updated.

malukenho commented on src/Middleware/ErrorResponseGenerator.php:
```diff
@@ -0,0 +1,145 @@
+     * @param ResponseInterface $response
+     * @return ResponseInterface
+     */
+    public function __invoke($e, ServerRequestInterface $request, ResponseInterface $response)
```
`\Exception $e`? I saw the `\Throwable` on the docblock, but still better rely on `\Exception` them nothing as it's not checking for it in the code.

weierophinney replied:
Except that `\Throwable` does not inherit from `\Exception` in PHP 7, which would make things like `TypeError` `$e` values create _new_ `TypeError`s.

This is a necessary evil until we can move to PHP 7+ only.

weierophinney replied:
Context is key here; it's being invoked from `Zend\Stratigility\Middleware\ErrorHandler`, which is invoking it with an exception or throwable that it has caught. Adding a check here is unnecessary.

malukenho commented on src/Middleware/NotFoundHandler.php:
```diff
@@ -0,0 +1,75 @@
+            return parent::process($request, $delegate);
+        }
+
+        $response = $this->responsePrototype->withStatus(404);
```
No constant for `404`?

weierophinney replied:
Not within our libraries, but it's a good point. I'll add `fig/http-message-util` as a dependency so we can use a constant here.

weierophinney commented:
Just did the following with [Adam Culp's Expressive workshop repo](https://github.com/adamculp/expressive-workshop) to see how upgrade might work:

- Added a repository entry to the `composer.json` pointing at my repository.
- Updated the `zend-expressive` requirement to `dev-feature/programmatic as 1.1.0`.
- Ran `composer update zendframework/zend-expressive zendframework/zend-stratigility` (the second package needed to be listed, as composer does not update nested dependencies when specifying a specific package).

At this point, the application continued to work, but emitted some deprecation notices due to the fact that, out-of-the-box, we use Stratigility error middleware.

I then ran `./vendor/bin/expressive-tooling` to install the zend-expressive-tooling package. Once installed, I ran `./vendor/bin/expressive-pipeline-from-config generate`. This ran without errors, and at this point, the application worked perfectly, emitting no errors.

I then ran each of the other migration tools (original messages and error middleware scanners); these caught nothing, meaning no further changes needed to be made.

I hand-edited the generated `config/pipeline.php` file, as the project defines a number of middleware that modify the returned request; I moved these to run before the error handler. Everything continues to work!

As such, I think we're in pretty good shape, but I'd like to see if anybody else can do similarly with their own apps and report back regarding the experience they had.

malukenho commented:
@weierophinney I did the upgrade as you said above, and worked fine for me.

danizord commented:
Worked fine here as well :+1:

xtreamwayz commented on bin/expressive-tooling:
```diff
@@ -0,0 +1,90 @@
+}
+
+$paths    = explode(PATH_SEPARATOR, $paths);
+$bins     = ['composer', 'composer.bat', 'composer.phar'];
```
Windows can have .cmd as well, which I have :)
```
$bins     = ['composer', 'composer.bat', 'composer.cmd', 'composer.phar'];
```

xtreamwayz commented:
I did this:
- Added a repository entry to the composer.json pointing at my repository.
- Updated the zend-expressive requirement to dev-feature/programmatic as 1.1.0.
Followed by `composer update` and it updated every package, including stratigility to version 1.3.1

I've converted 2 websites successfully.

webimpress commented on src/ApplicationConfigInjectionTrait.php:
```diff
@@ -0,0 +1,306 @@
+            $priority = isset($item['priority']) && is_int($item['priority'])
+                ? $item['priority']
+                : 1;
+            $queue->insert($item, [$priority, $serial--]);
```
We discuss before `++` operator and here it could be confusing.
I'd prefer to have here just `$serial` and then in new line `$serial--` (or `$serial -= 1`).
Ofc this is post decrementation so value is returned first and then value is changed.

weierophinney replied:
This code was originally copy-pasted from zend-stdlib's priority queue implementation (which ensures that items injected with the same priority execute in the order in which they are injected); likely need to revisit in that library as well. Updating now.

webimpress commented on test/ApplicationConfigInjectionTest.php:
```diff
@@ -0,0 +1,413 @@
+                }
+            }
+
+            if (! isset($spec['allowed_methods'])) {
```
`else` to above previous condition?

webimpress commented on src/Application.php:
```diff
@@ -426,7 +446,10 @@ public function routeMiddleware(ServerRequestInterface $request, ResponseInterfa
+
+                return $this->raiseThrowables
+                    ? $response
+                    : $next($request, $response, 405);
```
Should we use constant also here? (`StatusCodeInterface::STATUS_METHOD_NOT_ALLOWED`)

weierophinney replied:
Yep. Also needed to update TemplatedErrorHandler, as it was doing 404s.

weierophinney commented on src/ApplicationConfigInjectionTrait.php:
```diff
@@ -0,0 +1,306 @@
+                $methods = Route::HTTP_METHOD_ANY;
+            }
+            $name    = isset($spec['name']) ? $spec['name'] : null;
+            $route   = new Route($spec['path'], $spec['middleware'], $methods, $name);
```
Done.

weierophinney commented on src/ApplicationConfigInjectionTrait.php:
```diff
@@ -0,0 +1,306 @@
+     *         [
+     *             'path' => '/path/to/match',
+     *             'middleware' => 'Middleware Service Name or Callable',
+     *             'allowed_methods' => [ 'GET', 'POST', 'PATCH' ],
```
Done.

weierophinney commented on test/ApplicationConfigInjectionTest.php:
```diff
@@ -0,0 +1,413 @@
+           ['HelloWorld'],
+           [
+                function () {
+                }
```
Done

weierophinney commented on test/Container/ApplicationFactoryTest.php:
```diff
@@ -1308,4 +877,141 @@ public function testProperlyRegistersNestedErrorMiddlewareAsLazyErrorMiddleware(
+        $config = [
+            'middleware_pipeline' => [
+                [ 'path' => '/api', 'middleware' => $api ],
+                [ 'path' => '/dynamic-path', 'middleware' => 'DynamicPath' ],
```
Done, throughout (there were a lot of them!).

weierophinney commented on test/Container/NotFoundHandlerFactoryTest.php:
```diff
@@ -0,0 +1,72 @@
+        $this->container->get('config')->willReturn(['zend-expressive' => [
+            'error_handler' => [
+                'template_404' => 'error::custom404'
+            ]
```
Done.
